src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -163,146 +162,123 @@ public void Clear()
 
         /// <summary>
         /// Defines a cache of CachingContexts; instead of using a ConditionalWeakTable which can be slow to traverse
-        /// this approach uses a concurrent dictionary pointing to weak references of <see cref="CachingContext"/>.
-        /// Relevant caching contexts are looked up using the equality comparison defined by <see cref="EqualityComparer"/>.
+        /// this approach uses a fixed-size array of weak references of <see cref="CachingContext"/> that can be looked up lock-free.
+        /// Relevant caching contexts are looked up by linear traversal using the equality comparison defined by <see cref="EqualityComparer"/>.
         /// </summary>
         internal static class TrackedCachingContexts
         {
             private const int MaxTrackedContexts = 64;
-            private static readonly ConcurrentDictionary<JsonSerializerOptions, WeakReference<CachingContext>> s_cache =
-                new(concurrencyLevel: 1, capacity: MaxTrackedContexts, new EqualityComparer());
+            private static readonly WeakReference<CachingContext>?[] s_trackedContexts = new WeakReference<CachingContext>[MaxTrackedContexts];
+            private static int s_size;
 
-            private const int EvictionCountHistory = 16;
-            private static readonly Queue<int> s_recentEvictionCounts = new(EvictionCountHistory);
-            private static int s_evictionRunsToSkip;
+            private const int DanglingEntryEvictInterval = 8;
+            private static int s_lookupsWithDanglingEntries;
 
             public static CachingContext GetOrCreate(JsonSerializerOptions options)
             {
                 Debug.Assert(options.IsReadOnly, "Cannot create caching contexts for mutable JsonSerializerOptions instances");
                 Debug.Assert(options._typeInfoResolver != null);
 
-                ConcurrentDictionary<JsonSerializerOptions, WeakReference<CachingContext>> cache = s_cache;
+                if (TryGetSharedContext(options, out bool foundDanglingEntries, out CachingContext? result))
+                {
+                    // Periodically evict dangling entries in the case of successful lookups
+                    if (foundDanglingEntries && Interlocked.Increment(ref s_lookupsWithDanglingEntries) == DanglingEntryEvictInterval)
+                    {
+                        lock (s_trackedContexts)
+                        {
+                            EvictDanglingEntries();
```

Comment:
Is the whole purpose of this call to just clear the WeakReferences that got cleared?  I do not think you need to bother with that. Cleared WeakReferences cost very little. If the app peaked at certain number of caches entries, it is likely going to peak at the same number again - it is better to keep reusing the WeakReferences.

Reply:
It is also meant to push longer surviving caches towards the start of the array, so they can be found more quickly. Perhaps it's not too important though.